tools/testsys-config/src/lib.rs

```diff
@@ -0,0 +1,416 @@
+    /// The secrets needed by the agents
+    pub secrets: BTreeMap<String, SecretName>,
+    /// The role that should be assumed for this particular variant
+    pub assume_role: Option<String>,
+}
+
+/// Create a vec if relevant keys for this variant ordered from most specific to least specific.
+fn config_keys(variant: &Variant) -> Vec<String> {
+    let (family_flavor, platform_flavor) = variant
+        .variant_flavor()
+        .map(|flavor| {
+            (
+                format!("{}-{}", variant.family(), flavor),
+                format!("{}-{}", variant.platform(), flavor),
+            )
+        })
+        .unwrap_or_default();
+
+    // The keys used to describe configuration (most specific -> least specific)
+    vec![
+        variant.to_string(),
+        family_flavor,
+        variant.family().to_string(),
+        platform_flavor,
+        variant.platform().to_string(),
+    ]
+}
+
+#[derive(Debug, Default, Deserialize, Serialize, PartialEq, Eq, Clone)]
+#[serde(deny_unknown_fields)]
+pub struct GenericConfig {
+    #[serde(default)]
+    aarch64: GenericVariantConfig,
+    #[serde(default)]
+    x86_64: GenericVariantConfig,
+    #[serde(default, flatten)]
+    config: GenericVariantConfig,
+}
+
+impl GenericConfig {
+    /// Get the configuration for a specific arch.
+    pub fn for_arch<S1>(&self, arch: S1) -> GenericVariantConfig
+    where
+        S1: Into<String>,
+    {
+        match arch.into().as_str() {
+            "x86_64" => self.x86_64.clone(),
+            "aarch64" => self.aarch64.clone(),
+            _ => Default::default(),
+        }
+    }
+}
+
+#[derive(Debug, Default, Deserialize, Serialize, PartialEq, Eq, Clone)]
+#[serde(deny_unknown_fields)]
+#[serde(rename_all = "kebab-case")]
+pub struct GenericVariantConfig {
+    /// The names of all clusters this variant should be tested over. This is particularly useful
+    /// for testing Bottlerocket on ipv4 and ipv6 clusters.
+    #[serde(default)]
+    pub cluster_names: Vec<String>,
+    /// The instance type that instances should be launched with
+    pub instance_type: Option<String>,
+    /// The secrets needed by the agents
+    #[serde(default)]
+    pub secrets: BTreeMap<String, SecretName>,
+    /// The role that should be assumed for this particular variant
+    pub agent_role: Option<String>,
+    /// The custom images used for conformance testing
+    pub conformance_image: Option<String>,
+    /// The custom registry used for conformance testing
+    pub conformance_registry: Option<String>,
+}
+
+impl GenericVariantConfig {
+    /// Overwrite the unset values of `self` with the set values of `other`
+    fn merge(self, other: Self) -> Self {
+        let cluster_names = if self.cluster_names.is_empty() {
+            other.cluster_names
+        } else {
+            self.cluster_names
+        };
+
+        let secrets = if self.secrets.is_empty() {
+            other.secrets
+        } else {
+            self.secrets
+        };
```

Comment:
It is sort of tempting to have `cluster_names` and `secrets` be `Option`, and to have empty lists deserialize into `None`, just so all of the merge logic is the same. But I'm not sure it's worth it 😀